Post get paging

gil392 opened the pull request.

gil392 committed:
Using paging to load the posts. Added a unique path in the server. I keep state in usePosts.

gil392 committed:
Changed response from the server so I can know if there are more posts to get, now it's correct when there's none left. Previously it needed one empty request that returned nothing and then removed the refresh icon.

gil392 committed:
Merge branch 'development' of https://github.com/ItayFinkelstein/Frontend into post-get-paging

gil392 committed:
Fixed refresh button in posts page to look normal, and in the way also uses GenericIconButton now.

Not sure why we fetched posts after editing them, we just need to receive the data back from the edit request and change that post. I don't want to fetch 10 more posts because I edited an already shown post, and I certainly don't want to suddenly see the first 10 posts.

Fixed bugs from merge to make it work.

gil392 committed:
clean notes

ItayFinkelstein commented on src/PostPage.tsx:
```diff
@@ -26,7 +29,6 @@ export default function PostPage(props: PostPageProps) {
 
   const handlePostEdit = () => {
     setPostToEdit(null);
-    fetchPosts();
```
So it will not change? Can you check that after post edit it refreshes?

ItayFinkelstein commented on src/PostPage.tsx:
```diff
@@ -16,7 +19,7 @@ export type UserToDisplayProps = {
 
 export default function PostPage(props: PostPageProps) {
-  const { posts, fetchPosts } = usePosts();
+  const { posts, loadNextPage, isLoading, hasMore } = usePosts();
```
Please rename hasMore to hasMorePosts

ItayFinkelstein commented on src/data_hooks/usePosts.ts:
```diff
@@ -1,20 +1,56 @@
+  const [error, setError] = useState<Error | undefined>();
+  const [isLoading, setIsLoading] = useState(false);
+  const [currentPage, setCurrentPage] = useState(1);
+  const [hasMore, setHasMore] = useState(true);
```
Rename hasMore to hasMorePosts.

ItayFinkelstein commented on src/data_hooks/usePosts.ts:
```diff
@@ -1,20 +1,56 @@
+  const [posts, setPosts] = useState<Post[]>([]);
+  const [error, setError] = useState<Error | undefined>();
+  const [isLoading, setIsLoading] = useState(false);
+  const [currentPage, setCurrentPage] = useState(1);
```
Please have this number in config

gil392 replied:
That makes no sense. You always start to fetch from the first data you have. It means "start fetching from the first batch of posts", so it's always 1.

ItayFinkelstein commented on src/data_hooks/usePosts.ts:
```diff
@@ -1,20 +1,56 @@
+        setIsLoading(false);
+      })
+      .catch((err) => {
+        if (err instanceof CanceledError) return; // A "good" error, it means something decides to cancel the fetch request.
```
Something? Like what?

gil392 replied:
For example, when we want to refetch and then you'd want to cancel the previous fetch request. It didn't fail, we just wanted a more recent version before it got completed.

ItayFinkelstein commented on src/data_hooks/usePosts.ts:
```diff
@@ -1,20 +1,56 @@
+
+  const loadNextPage = () => {
+    if (hasMore && !isLoading) {
+      fetchPosts(currentPage + 1);
```
Is this plus 1 even if the fetch count is more than 1?

gil392 replied:
Yes, imagine it's like google searches when you move to the next page of results. "Page" means the next batch of posts.
If you prefer, I can rename page to something else, I'm just not sure to what.

ItayFinkelstein reviewed (changes requested):
Please check my comments

gil392 committed:
Update post- when we get the response from the server, update it in the posts list.

Had a bug where I couldn't do it in PostCardForm, because I edited the array that created it, so setPosts never happened.

gil392 committed:
PR Fixes: renaming

gil392 committed:
Trying to set filter func so between users it's fine

gil392 committed:
try to fix the problem
